propagation/http_trace_context_propagator_test.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation_test
+
+import (
+	"context"
+	"net/http"
+	"testing"
+
+	"go.opentelemetry.io/api/trace"
+
+	"github.com/google/go-cmp/cmp"
+
+	"go.opentelemetry.io/api/core"
+	mocktrace "go.opentelemetry.io/internal/trace"
+	"go.opentelemetry.io/propagation"
+)
+
+var (
+	traceID = core.TraceID{High: 0x4bf92f3577b34da6, Low: 0xa3ce929d0e0e4736}
+	spanID  = uint64(0x00f067aa0ba902b7)
+)
+
+func TestExtractTraceContextFromHTTPReq(t *testing.T) {
+	trace.SetGlobalTracer(&mocktrace.MockTracer{})
+	propagator := propagation.HttpTraceContextPropagator()
+	tests := []struct {
+		name   string
+		header string
+		wantSc core.SpanContext
+	}{
+		{
+			name:   "future version",
+			header: "02-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01",
+			wantSc: core.SpanContext{
+				TraceID:      traceID,
+				SpanID:       spanID,
+				TraceOptions: core.TraceOptionSampled,
+			},
+		},
+		{
+			name:   "zero trace ID and span ID",
+			header: "00-00000000000000000000000000000000-0000000000000000-01",
+			wantSc: core.EmptySpanContext(),
+		},
+		{
+			name:   "valid header",
+			header: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01",
+			wantSc: core.SpanContext{
+				TraceID:      traceID,
+				SpanID:       spanID,
+				TraceOptions: core.TraceOptionSampled,
+			},
+		},
+		{
+			name:   "missing options",
+			header: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7",
+			wantSc: core.EmptySpanContext(),
+		},
+		{
```

Comment:
Some suggestions for the tests:

- "wrong length of version", "0000-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"
- "bogus version", "qw-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"
- "invalid version", "ff-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"
- "wrong length of trace id", "00-4bf92f3577b34da6a3ce929d0e0e47-00f067aa0ba902b7-01"
- "bogus trace id", "00-qwf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"
- "wrong length of parent id", "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902-01"
- "bogus parent id", "00-4bf92f3577b34da6a3ce929d0e0e4736-qwf067aa0ba902b7-01"
- "valid trace id, zero parent id", "00-4bf92f3577b34da6a3ce929d0e0e4736-0000000000000000-01"
- "valid parent id, zero trace id", "00-00000000000000000000000000000000-00f067aa0ba902b7-01"
- "wrong length of options", "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-0101"
- "invalid options", "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-02"
- "bogus options", "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-qw"
- "ignored future options", "aa-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-ff"
  - should result in valid span context with only sampling bit set in options

Not sure about those yet:

- "version with invalid hex characters", "0A-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"
- "trace id with invalid hex characters", "00-4BF92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01"
- "parent id with invalid hex characters", "00-4BF92f3577b34da6a3ce929d0e0e4736-00F067AA0ba902b7-01"
- "options with invalid hex characters", "01-4BF92f3577b34da6a3ce929d0e0e4736-00F067AA0ba902b7-A1"

Reply:
added all test cases including upper-case hex characters.
Also added future additional data "aa-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-ff-XYZxsf09"